text/0000-async-signal-safe-thread-local.md

```diff
@@ -0,0 +1,315 @@
+    });
+}
+```
+
+With optimization enabled, this translates to the following pseudo-assembly:
+
+```
+main:
+
+ 1:  if X is not initialized:
+ 2:      initialize X
+ 3:  endif
+ 4:  if x is borrowed:
+ 5:      panic
+ 6:  endif
+ 7:  increase x.0 by 1
+ 8:  increase x.1 by 1
+```
+
+The missing step is after `6`. The compiler does not emit an instruction to mark
+`x` as borrowed. With this instruction, it would look like this:
+
+```
+ 6:  endif
+ 7:  mark x as borrowed
+ 8:  increase x.0 by 1
+ 9:  increase x.1 by 1
+10:  mark x as not borrowed
+```
+
+The compiler does not emit these instructions because it does not believe that
+the difference can be observed. However, consider what happens when the
+following signal handler is invoked after operation `7` but before operation `8`
+in the original listing:
+
+```rust
+extern fn handler(_: i32) {
+    X.with(|x| {
+        let x = x.borrow();
+        println!("{:?}", *x);
+    });
+}
+```
+
+Since `x` was never marked as borrowed, the borrow succeeds and `(1, 0)` is
+printed.
+
+This shows that, in its current state, rust code must not access thread local
+variables declared with `thread_local!` inside of signal handlers.
+
+# Detailed design
+[design]: #detailed-design
+
+## Changes to the language
+
+A new language item and marker trait `Interrupt` is added to the language:
+
+```rust
+/// Types that can be safely accessed from signal handlers.
+///
+/// The precise definition is: a type `T` is `Interrupt` if `&T` is
+/// async-signal-safe. In other words, there is no possibility of data
+/// inconsistency when `&T` is used inside a signal handler.
+///
+/// [...]
+#[lang = "interrupt"]
+pub unsafe trait Interrupt {
+    // Empty
+}
+
+unsafe impl Interrupt for .. { }
+```
+
+The types that are excluded from `Sync` by default are also excluded from
+`Interrupt`:
+
+```rust
+impl<T> !Interrupt for *const T { }
+impl<T> !Interrupt for *mut T { }
+impl<T> !Interrupt for UnsafeCell<T> { }
+// etc.
+```
+
+All types that are `Sync` are also `Interrupt`:
+
+```rust
+unsafe impl<T: Sync> Interrupt for T { }
+```
+
+Statics marked with `#[thread_local]` will accept exactly those types that
+implement the `Interrupt` trait.
+
+## Changes to the standard library
+
+`thread_local!` is changed to accept only `Interrupt` types. This is a breaking
+change, however, the author believes that it is easily mitigated for the
+following reasons:
+
+1. Thread local storage is only useful with (interior) mutability.
+2. Thread local storage created with `thread_local!` only allows interior
+   mutability.
+3. The types most commonly used for interior mutability are `Cell`, `RefCell`,
+   and `Sync` types with interior mutability.
+
+Hence, the author believes that almost all cases of broken code will be caused
+by either `Cell` or `RefCell`.
+
+`RefCell` can be fixed easily: Since there already is a locking mechanism, one
+only has to ensure that the locking instructions will actually be emitted. Since
```

Comment:
>It's not obvious that property will still hold for #[thread_local], though.

Without a `'thread` lifetime `#[thread_local]` variables are already unsafe because they pretend to be `'static`.

>Even if we don't need an atomic increment, I'm not sure it's appropriate to completely ignore the performance cost of the optimization barrier.

It's certainly easy to write another type that replaces `RefCell` if this is true. In this case the fix would be to replace the name `RefCell` by the name of its signal-safe replacement.